File: ios-native/BuilderHQ/DesignSystem/CelebrationScene.swift
/// CelebrationScene — the shared, choreographed "moment" overlay.
///
/// One premium celebration system for every achievement in the app —
/// project published, builder unlocked, tender sent, tender awarded —
/// so they feel like one coherent, intentional language instead of
/// three hand-rolled one-offs.
///
/// The choreography (Strava-grade, but elegant/restrained):
///
///   stage 0  backdrop frosts in
///   stage 1  medallion springs in (0.7→1) + the ring begins to draw
///            itself around it  ·  light haptic
///   stage 2  ring completes, sonar ripples ping outward, icon bounces,
///            particles rise  ·  medium haptic
///   stage 3  headline rises + fades in
///   stage 4  detail / subtitle / metric resolve  ·  success haptic
///   stage 5  CTA (if any) rises in
///
/// The signature element is the **self-drawing progress ring** — an
/// angular teal→blue gradient stroke that sweeps closed like an Apple
/// activity ring completing. That single move is what makes it read
/// as "premium" rather than "a popup".
///
/// The scene runs its own entrance on `.onAppear`; the *parent* owns
/// when it's shown/hidden and any auto-dismiss timing (unchanged from
/// before). Pass a `cta` for a full-screen moment with a bottom
/// button; omit it for a centered auto-dismissing moment.

import SwiftUI

// MARK: - Tone

enum CelebrationTone {
    case accent   // teal — the everyday win (unlock, tender sent)
    case success  // brighter — project went live
    case gold     // the peak — tender awarded

    var primary: Color {
        switch self {
        case .accent:  return Palette.accent
        case .success: return Palette.accentLight
        case .gold:    return Color(hex: 0xF5C451)
        }
    }
    var secondary: Color {
        switch self {
        case .accent:  return Palette.blue
        case .success: return Palette.accent
        case .gold:    return Color(hex: 0xE08A2B)
        }
    }
    var glow: Color { primary.opacity(0.5) }
    var contrast: Color { Palette.accentContrast }
}

// MARK: - CTA

struct CelebrationCTA {
    let label: String
    let action: () -> Void
}

// MARK: - Scene

struct CelebrationScene: View {
    let symbol: String
    var tone: CelebrationTone = .accent
    /// Plain leading words of the headline (e.g. "Your project is").
    /// May be empty for a single-word headline.
    var titlePlain: String = ""
    /// The emphasis word, rendered in serif italic accent ("live.").
    let titleEmphasis: String
    /// Optional strong detail line (e.g. the project title).
    var detail: String? = nil
    var subtitle: String? = nil
    /// Optional metric pill (e.g. "You saved $249").
    var metric: String? = nil
    var metricIcon: String = "sparkles"
    var cta: CelebrationCTA? = nil

    /// Drives the staged choreography. Advanced on appear.
    @State private var stage: Int = 0
    /// 0→1 sweep that draws the ring closed.
    @State private var ringProgress: CGFloat = 0
    /// Toggles the continuous sonar ripple + medallion breathing.
    @State private var alive: Bool = false

    var body: some View {
        ZStack {
            backdrop

            if cta != nil {
                // Full-screen moment with a bottom CTA.
                VStack(spacing: 0) {
                    Spacer()
                    centerpiece
                    textBlock.padding(.top, 36)
                    Spacer()
                    Spacer()
                    ctaButton
                }
            } else {
                // Centered, auto-dismissing moment.
                VStack(spacing: 28) {
                    centerpiece
                    textBlock
                }
            }
        }
        .sensoryFeedback(trigger: stage) { _, s in
            switch s {
            case 1:  return .impact(weight: .light)
            case 2:  return .impact(weight: .medium)
            case 4:  return .success
            default: return nil
            }
        }
        .onAppear(perform: runChoreography)
    }

    // MARK: - Backdrop

    private var backdrop: some View {
        Rectangle()
            .fill(.ultraThinMaterial)
            .ignoresSafeArea()
            .overlay(Palette.canvas.opacity(0.55).ignoresSafeArea())
            .opacity(stage >= 0 ? 1 : 0)
    }

    // MARK: - Centerpiece (medallion + ring + ripples + particles)

    private var centerpiece: some View {
        ZStack {
            // Particle field rising behind the medallion.
            CelebrationSparkles(tone: tone, fire: alive)
                .frame(width: 300, height: 300)

            // Sonar ripples — three rings pinging outward. Technical /
            // radar feel that ties to the blueprint identity.
            ForEach(0..<3, id: \.self) { i in
                Circle()
                    .stroke(tone.primary.opacity(0.26), lineWidth: 1.1)
                    .frame(width: 116, height: 116)
                    .scaleEffect(alive ? 2.3 : 0.92)
                    .opacity(alive ? 0 : 0.7)
                    .animation(
                        .easeOut(duration: 1.9)
                            .repeatForever(autoreverses: false)
                            .delay(Double(i) * 0.5),
                        value: alive
                    )
            }

            // Soft radial core glow.
            Circle()
                .fill(
                    RadialGradient(
                        colors: [tone.primary.opacity(0.30), tone.primary.opacity(0.04)],
                        center: .center,
                        startRadius: 6,
                        endRadius: 80
                    )
                )
                .frame(width: 150, height: 150)
                .shadow(color: tone.glow, radius: 40)
                .scaleEffect(stage >= 1 ? 1 : 0.7)
                .opacity(stage >= 1 ? 1 : 0)

            // The self-drawing progress ring — the signature move.
            Circle()
                .trim(from: 0, to: ringProgress)
                .stroke(
                    AngularGradient(
                        colors: [
                            tone.primary, tone.secondary, tone.primary,
                        ],
                        center: .center
                    ),
                    style: StrokeStyle(lineWidth: 3, lineCap: .round)
                )
                .rotationEffect(.degrees(-90))
                .frame(width: 132, height: 132)
                .shadow(color: tone.glow, radius: 10)

            // The icon.
            Image(systemName: symbol)
                .font(.system(size: 46, weight: .bold))
                .foregroundStyle(tone.primary)
                .shadow(color: tone.glow, radius: 18)
                .scaleEffect(stage >= 1 ? (alive ? 1.0 : 0.5) : 0.5)
                .opacity(stage >= 1 ? 1 : 0)
                .symbolEffect(.bounce, value: stage >= 2)
                .animation(.spring(response: 0.5, dampingFraction: 0.55), value: stage >= 1)
        }
        .frame(width: 200, height: 200)
    }

    // MARK: - Text

    private var textBlock: some View {
        VStack(spacing: 12) {
            // Headline — plain words + serif-italic emphasis word.
            VStack(spacing: -6) {
                if !titlePlain.isEmpty {
                    Text(titlePlain)
                        .font(Typography.ui(size: 27, weight: .medium))
                        .foregroundStyle(Palette.text)
                        .tracking(-0.4)
                }
                Text(titleEmphasis)
                    .font(Typography.serifItalic(size: titlePlain.isEmpty ? 40 : 38))
                    .foregroundStyle(tone.primary == Palette.accent ? Palette.accentLight : tone.primary)
                    .tracking(-0.4)
            }
            .opacity(stage >= 3 ? 1 : 0)
            .offset(y: stage >= 3 ? 0 : 14)

            if let detail {
                Text(detail)
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundStyle(Palette.textMuted)
                    .multilineTextAlignment(.center)
                    .padding(.top, 2)
                    .opacity(stage >= 4 ? 1 : 0)
                    .offset(y: stage >= 4 ? 0 : 10)
            }

            if let subtitle {
                Text(subtitle)
                    .font(.system(size: 13, weight: .regular))
                    .foregroundStyle(Palette.textDim)
                    .multilineTextAlignment(.center)
                    .fixedSize(horizontal: false, vertical: true)
                    .padding(.horizontal, 40)
                    .opacity(stage >= 4 ? 1 : 0)
                    .offset(y: stage >= 4 ? 0 : 10)
            }

            if let metric {
                HStack(spacing: 6) {
                    Image(systemName: metricIcon)
                        .font(.system(size: 11, weight: .bold))
                    Text(metric)
                        .font(.system(size: 12, weight: .bold))
                        .tracking(0.2)
                }
                .foregroundStyle(tone.primary)
                .padding(.horizontal, 12)
                .padding(.vertical, 7)
                .background(Capsule().fill(tone.primary.opacity(0.12)))
                .overlay(Capsule().stroke(tone.primary.opacity(0.35), lineWidth: 1))
                .padding(.top, 6)
                .opacity(stage >= 4 ? 1 : 0)
                .scaleEffect(stage >= 4 ? 1 : 0.9)
            }
        }
        .animation(.spring(response: 0.55, dampingFraction: 0.86), value: stage)
    }

    // MARK: - CTA

    @ViewBuilder
    private var ctaButton: some View {
        if let cta {
            Button(action: cta.action) {
                HStack(spacing: 10) {
                    Text(cta.label)
                        .font(.system(size: 16, weight: .bold))
                    Image(systemName: "arrow.right")
                        .font(.system(size: 14, weight: .bold))
                }
                .foregroundStyle(tone.contrast)
                .frame(height: 56)
                .frame(maxWidth: .infinity)
                .background(Capsule().fill(tone.primary))
                .shadow(color: tone.glow, radius: 22, x: 0, y: 10)
            }
            .padding(.horizontal, 24)
            .padding(.bottom, 32)
            .opacity(stage >= 5 ? 1 : 0)
            .offset(y: stage >= 5 ? 0 : 14)
            .animation(.spring(response: 0.55, dampingFraction: 0.86), value: stage >= 5)
        }
    }

    // MARK: - Choreography

    private func runChoreography() {
        // Backdrop is up at stage 0 immediately.
        advance(to: 1, after: 0.10) {
            withAnimation(.easeOut(duration: 0.95)) { ringProgress = 1 }
        }
        advance(to: 2, after: 0.34) {
            withAnimation(.easeOut(duration: 0.2)) { alive = true }
        }
        advance(to: 3, after: 0.50)
        advance(to: 4, after: 0.70)
        advance(to: 5, after: 0.92)
    }

    private func advance(to value: Int, after delay: Double, _ extra: (() -> Void)? = nil) {
        DispatchQueue.main.asyncAfter(deadline: .now() + delay) {
            stage = value
            extra?()
        }
    }
}

// MARK: - Particle field

/// Refined sparkle field — small tone-tinted dots that rise and fade
/// once on `fire`. Varied size / drift / lifetime so it reads as a
/// gentle shimmer, never a confetti dump.
private struct CelebrationSparkles: View {
    let tone: CelebrationTone
    let fire: Bool

    private struct Particle: Identifiable {
        let id: Int
        let x: CGFloat
        let yStart: CGFloat
        let size: CGFloat
        let delay: Double
        let duration: Double
        let useSecondary: Bool
    }

    // Deterministic spread (no Math.random at runtime — varied by index).
    private let particles: [Particle] = (0..<16).map { i in
        let golden = Double(i) * 2.399963   // golden-angle scatter
        let x = CGFloat(cos(golden)) * CGFloat(40 + (i % 5) * 22)
        return Particle(
            id: i,
            x: x,
            yStart: CGFloat((i * 37) % 60) - 10,
            size: CGFloat(3 + (i % 3)),
            delay: Double(i % 6) * 0.05,
            duration: 1.7 + Double(i % 4) * 0.25,
            useSecondary: i % 3 == 0
        )
    }

    var body: some View {
        GeometryReader { geo in
            ZStack {
                ForEach(particles) { p in
                    Circle()
                        .fill(p.useSecondary ? tone.secondary : tone.primary)
                        .frame(width: p.size, height: p.size)
                        .shadow(color: tone.glow, radius: 5)
                        .position(
                            x: geo.size.width / 2 + p.x,
                            y: fire
                                ? geo.size.height * 0.18
                                : geo.size.height * 0.56 + p.yStart
                        )
                        .opacity(fire ? 0 : 0.9)
                        .animation(
                            .easeOut(duration: p.duration).delay(p.delay),
                            value: fire
                        )
                }
            }
        }
    }
}

File: android-native/app/src/main/java/au/com/builderhq/app/feature/owner/CelebrationScene.kt
package au.com.builderhq.app.feature.owner

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.slideInVertically
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import au.com.builderhq.app.core.design.components.PrimaryButton
import au.com.builderhq.app.core.design.rememberHaptics
import au.com.builderhq.app.core.design.theme.Bhq
import au.com.builderhq.app.core.design.theme.DisplaySerif
import au.com.builderhq.app.core.design.theme.Motion
import kotlinx.coroutines.delay
import kotlin.math.cos
import kotlin.math.sin

enum class CelebrationTone { ACCENT, SUCCESS, GOLD }

/**
 * Full-screen achievement moment. Staged choreography: medallion springs in,
 * a teal ring self-draws, sonar ripples ping outward, a golden-angle particle
 * burst rises, then headline → detail → CTA resolve. Shared by publish + award.
 */
@Composable
fun CelebrationScene(
    icon: ImageVector,
    titlePlain: String,
    titleEmphasis: String,
    detail: String,
    subtitle: String,
    metric: String?,
    ctaLabel: String,
    tone: CelebrationTone,
    onCta: () -> Unit,
) {
    val c = Bhq.colors
    val haptics = rememberHaptics()
    var stage by remember { mutableIntStateOf(0) }
    LaunchedEffect(Unit) {
        haptics.tick()
        delay(120); stage = 1
        delay(260); stage = 2; haptics.press()
        delay(200); stage = 3
        delay(220); stage = 4; haptics.confirm()
        delay(240); stage = 5
    }
    val color = when (tone) {
        CelebrationTone.ACCENT -> c.accent
        CelebrationTone.SUCCESS -> c.success
        CelebrationTone.GOLD -> c.gold
    }
    val ring by animateFloatAsState(if (stage >= 1) 1f else 0f, tween(950, easing = Motion.EaseOutSoft), label = "ring")
    val medScale by animateFloatAsState(if (stage >= 1) 1f else 0.5f, spring(dampingRatio = 0.5f, stiffness = 520f), label = "med")
    val burst by animateFloatAsState(if (stage >= 2) 1f else 0f, tween(900, easing = Motion.EaseOut), label = "burst")
    val infinite = rememberInfiniteTransition(label = "ripple")
    val rip by infinite.animateFloat(0f, 1f, infiniteRepeatable(tween(2200, easing = LinearEasing)), label = "rip")

    Box(Modifier.fillMaxSize().background(c.scrimHeavy), contentAlignment = Alignment.Center) {
        Column(
            Modifier.padding(36.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Box(Modifier.size(160.dp), contentAlignment = Alignment.Center) {
                if (stage >= 2) {
                    Canvas(Modifier.fillMaxSize()) {
                        repeat(3) { i ->
                            val p = (rip + i / 3f) % 1f
                            val r = (size.minDimension / 2f) * (0.45f + p * 1.4f)
                            drawCircle(color.copy(alpha = (1f - p) * 0.32f), r, style = Stroke(2.dp.toPx()))
                        }
                    }
                    Canvas(Modifier.fillMaxSize()) {
                        repeat(16) { i ->
                            val ang = i * 2.399963f // golden angle
                            val dist = size.minDimension * (0.16f + burst * 0.42f)
                            val pos = Offset(center.x + cos(ang) * dist, center.y + sin(ang) * dist)
                            drawCircle(color.copy(alpha = (1f - burst) * 0.9f), 2.5.dp.toPx(), pos)
                        }
                    }
                }
                Canvas(Modifier.size(116.dp)) {
                    val sw = 4.dp.toPx()
                    val box = Size(size.width - sw, size.height - sw)
                    val tl = Offset(sw / 2, sw / 2)
                    drawArc(color.copy(0.15f), -90f, 360f, false, topLeft = tl, size = box, style = Stroke(sw))
                    drawArc(
                        brush = Brush.sweepGradient(listOf(color, c.accentLight, color)),
                        startAngle = -90f, sweepAngle = 360f * ring, useCenter = false,
                        topLeft = tl, size = box, style = Stroke(sw, cap = StrokeCap.Round),
                    )
                }
                Box(
                    Modifier
                        .size(80.dp)
                        .graphicsLayer { scaleX = medScale; scaleY = medScale }
                        .clip(RoundedCornerShape(50))
                        .background(Brush.verticalGradient(listOf(c.accentLight, color))),
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(icon, contentDescription = null, tint = c.accentContrast, modifier = Modifier.size(40.dp))
                }
            }

            Spacer(Modifier.height(28.dp))
            AnimatedVisibility(stage >= 3, enter = fadeIn() + slideInVertically { it / 3 }) {
                Text(
                    buildAnnotatedString {
                        append("$titlePlain ")
                        withStyle(SpanStyle(color = color, fontFamily = DisplaySerif, fontStyle = FontStyle.Italic, fontWeight = FontWeight.Normal)) {
                            append(titleEmphasis)
                        }
                    },
                    color = c.text, fontSize = 30.sp, fontWeight = FontWeight.Bold,
                    textAlign = TextAlign.Center,
                )
            }
            Spacer(Modifier.height(10.dp))
            AnimatedVisibility(stage >= 4, enter = fadeIn() + slideInVertically { it / 4 }) {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(detail, color = color, fontSize = 16.sp, fontWeight = FontWeight.SemiBold, textAlign = TextAlign.Center)
                    Spacer(Modifier.height(8.dp))
                    Text(subtitle, color = c.textMuted, fontSize = 14.sp, lineHeight = 21.sp, textAlign = TextAlign.Center)
                    if (metric != null) {
                        Spacer(Modifier.height(14.dp))
                        Box(
                            Modifier
                                .clip(RoundedCornerShape(50))
                                .background(color.copy(alpha = 0.12f))
                                .border(1.dp, color.copy(alpha = 0.35f), RoundedCornerShape(50))
                                .padding(horizontal = 14.dp, vertical = 7.dp),
                        ) {
                            Text(metric, color = color, fontSize = 13.sp, fontWeight = FontWeight.SemiBold)
                        }
                    }
                }
            }
            Spacer(Modifier.height(30.dp))
            AnimatedVisibility(stage >= 5, enter = fadeIn() + slideInVertically { it / 3 }) {
                PrimaryButton(ctaLabel, onClick = onCta)
            }
        }
    }
}
